src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -0,0 +1,392 @@
+                _ptr,
+                (ushort)_remoteEndPoint.AddressFamily,
+                _remoteEndPoint.Address.ToString(),
+                (ushort)_remoteEndPoint.Port));
+
+            return _connectTcs.GetTypelessValueTask();
+        }
+
+        private MsQuicStream StreamOpen(
+            QUIC_STREAM_OPEN_FLAG flags)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            IntPtr streamPtr = IntPtr.Zero;
+            MsQuicStatusException.ThrowIfFailed(
+                MsQuicApi.Api.StreamOpenDelegate(
+                _ptr,
+                (uint)flags,
+                MsQuicStream.NativeCallbackHandler,
+                IntPtr.Zero,
+                out streamPtr));
+
+            MsQuicStream stream = new MsQuicStream(this, flags, streamPtr, inbound: false);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+            return stream;
+        }
+
+        private void SetCallbackHandler()
+        {
+            _handle = GCHandle.Alloc(this);
+            _connectionDelegate = new ConnectionCallbackDelegate(NativeCallbackHandler);
+            MsQuicApi.Api.SetCallbackHandlerDelegate(
+                _ptr,
+                _connectionDelegate,
+                GCHandle.ToIntPtr(_handle));
+        }
+
+        private ValueTask ShutdownAsync(
+            QUIC_CONNECTION_SHUTDOWN_FLAG Flags,
+            ushort ErrorCode)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            uint status = MsQuicApi.Api.ConnectionShutdownDelegate(
+                _ptr,
+                (uint)Flags,
+                ErrorCode);
+            MsQuicStatusException.ThrowIfFailed(status);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+            return _shutdownTcs.GetTypelessValueTask();
+        }
+
+        internal static uint NativeCallbackHandler(
+            IntPtr connection,
+            IntPtr context,
+            ref ConnectionEvent connectionEventStruct)
+        {
+            GCHandle handle = GCHandle.FromIntPtr(context);
+            MsQuicConnection quicConnection = (MsQuicConnection)handle.Target;
+            return quicConnection.HandleEvent(ref connectionEventStruct);
+        }
+
+        public override void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        ~MsQuicConnection()
+        {
+            Dispose(false);
+        }
+
+        private void Dispose(bool disposing)
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            if (_ptr != IntPtr.Zero)
+            {
+                MsQuicApi.Api.ConnectionCloseDelegate?.Invoke(_ptr);
+            }
+
+            _ptr = IntPtr.Zero;
+
+            _handle.Free();
+            _session?.Dispose();
+            _securityConfig?.Dispose();
```

Comment:
It's not safe to call these if `disposing == false`. Use ref-counting `SafeHandle` if needed.

Reply:
I'm definitely leaving converting to safehandles as a TODO <_<.